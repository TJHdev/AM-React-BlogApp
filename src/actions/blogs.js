import database from '../firebase/firebase.js'

// component calls action generator
// action generator returns object
// component dispatches object
// redux store changes

// component calls action generator
// action generator returns function
// component dispatches object (?)
// function runs (has the ability to dispatch other actions and do whatever it wants)

export const addBlog = (blog) => ({
  type: 'ADD_BLOG',
  blog: blog
})

export const startAddBlog = (blogData = {}) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    const {
      title = '',
      body = '',
      createdAt = 0,
      lastEditedAt = 0
    } = blogData;
    
    const blog = { title, body, createdAt, lastEditedAt };

    return database.ref(`users/${uid}/blogs`).push(blog).then((ref) => {
      dispatch(addBlog({
        id: ref.key,
        ...blog
      }));
    });
  };
};

export const removeBlog = ({ id } = {}) => ({
  type: 'REMOVE_BLOG',
  id: id
});

export const startRemoveBlog = ({ id }) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database.ref(`users/${uid}/blogs/${id}`).set(null).then(() => {
      dispatch(removeBlog({ id }));
    });
  }
};

export const editBlog = (id, updates) => ({
  type: 'EDIT_BLOG',
  id: id,
  updates
});

export const startEditBlog = (id, updates = {}) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database.ref(`users/${uid}/blogs/${id}`).update(updates).then(() => {
      dispatch(editBlog(id, updates));
    }); 
  };
};

export const setBlogs = (blogs) => ({
  type: 'SET_BLOGS',
  blogs: blogs
});

export const startSetBlogs = (blogsData = []) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database.ref(`users/${uid}/blogs`).once('value').then((snapshot) => {
      const blogs = [];
      snapshot.forEach((childSnapshot) => {
        blogs.push({
          id: childSnapshot.key,
          ...childSnapshot.val()
        })
      })
      dispatch(setBlogs(blogs));
    });
  }
};

export const startSetBlogsPublic = (blogsData = []) => {
  return (dispatch, getState) => {
    return database.ref('users').once('value').then((snapshot) => {
      const blogs = [];
      const val = snapshot.val();

      Object.keys(val).forEach((childSnapshot) => {
        Object.keys(val[childSnapshot].blogs).forEach((blog) => {
          blogs.push({
            id: blog,
            ...snapshot.val()[childSnapshot].blogs[blog]
          })
        })
      })
      dispatch(setBlogs(blogs));
    });
  }
};

