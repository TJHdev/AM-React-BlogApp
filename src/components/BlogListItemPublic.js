import React from 'react';
import { Link } from 'react-router-dom';
import moment from 'moment';

const BlogListItemPublic = ({ id, title, createdAt }) => (
  <Link className="list-item" to={`/read/${id}`}>
    <h3 className="list-item__title">{title}</h3>
    <span className="list-item__subtitle">{moment(createdAt).format('h:mma, MMMM Do, YYYY')}</span>
  </Link>
);

export default BlogListItemPublic;