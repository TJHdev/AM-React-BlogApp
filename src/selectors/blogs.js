import moment from 'moment';

// Get visible blogs

const getVisibleBlogs = (blogs, { text, sortBy }) => {
  return blogs.filter((blog) => {
    const textMatch = blog.title.toLowerCase().includes(text.toLowerCase());
    return textMatch;
  }).sort((a, b) => {
    if (sortBy === 'dateCreated') {
      return a.createdAt < b.createdAt ? 1 : -1;
    } else if (sortBy === 'dateEdited') {
      return a.lastEditedAt < b.lastEditedAt ? 1 : -1;
    } else if (sortBy === 'title') {
      return a.title < b.title ? 1 : -1;
    }
  });
};

export default getVisibleBlogs;