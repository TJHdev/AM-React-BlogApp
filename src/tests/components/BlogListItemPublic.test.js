import React from 'react';
import { shallow } from 'enzyme';
import BlogListItemPublic from '../../components/BlogListItemPublic';
import blogs from '../fixtures/blogs';

test('should render public Blog list item correctly', () => {
  const wrapper = shallow(<BlogListItemPublic {...blogs[0]} />);
  expect(wrapper).toMatchSnapshot();
});