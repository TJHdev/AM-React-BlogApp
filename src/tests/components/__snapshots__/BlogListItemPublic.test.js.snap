// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render public Blog list item correctly 1`] = `
<Link
  className="list-item"
  replace={false}
  to="/read/1"
>
  <h3
    className="list-item__title"
  >
    mechanical engineering sucks
  </h3>
  <span
    className="list-item__subtitle"
  >
    1:00am, January 1st, 1970
  </span>
</Link>
`;
